skyd/skyd.go

```diff
@@ -1,16 +1,149 @@
 package skyd
 
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	skydclient "gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+var (
+	// skylinksCache is a local cache of the list of skylinks pinned by the
+	// local skyd
+	skylinksCache pinnedSkylinksCache
+	// skylinksCacheDuration defines the duration of skylinksCache
+	skylinksCacheDuration = time.Hour
+)
+
 type (
-	// Client allows us to call the local skyd instance.
-	Client struct {
+	// Client describes the interface exposed by client.
+	Client interface {
+		Pin(skylink string) error
+		PinnedSkylinks() (skylinks map[string]interface{}, err error)
+		Unpin(skylink string) error
+	}
+
+	// client allows us to call the local skyd instance.
+	client struct {
+		staticClient *skydclient.Client
+	}
+	// renterDirCache is a simple cache of the renter's directory information,
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   map[string]interface{}
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) Client {
+	opts := skydclient.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &client{
+		staticClient: skydclient.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: skymodules.RandomSiaPath(),
+	}
+	err = c.staticClient.SkynetSkylinkPinPost(skylink, spp)
+	if err != nil {
+		c.updateCachedStatus(skylink, false)
+	}
+	return err
+}
+
+// PinnedSkylinks returns the list of skylinks pinned by the local skyd.
+func (c *client) PinnedSkylinks() (map[string]interface{}, error) {
+	skylinksCache.mu.Lock()
+	defer skylinksCache.mu.Unlock()
+	// Check whether the cache is still valid and return it if so.
+	if skylinksCache.Expiration.After(time.Now().UTC()) {
+		return skylinksCache.Skylinks, nil
+	}
+	// The cache is not valid, fetch the data from skyd.
+	sls := make(map[string]interface{})
+	dirsToWalk := []skymodules.SiaPath{skymodules.SkynetFolder}
+	for len(dirsToWalk) > 0 {
+		// Pop the first dir and walk it.
+		dir := dirsToWalk[0]
+		dirsToWalk = dirsToWalk[1:]
+
+		rd, err := c.staticClient.RenterDirRootGet(dir)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to fetch skynet directories from skyd")
```

Comment:
I'm wondering whether out of a safety precaution we shouldn't be logging a critical here but continue our walk?

Reply:
This whole thing will be replaced by a native skyd call, so i wouldn't worry about it too much.  
I can add a critical, if you think it would be better - I don't mind.

Reply:
recommend ignoring this case as this will all be moved to skyd and become obsolete.